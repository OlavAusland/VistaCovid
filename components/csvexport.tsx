import { collection, getDocs } from 'firebase/firestore';
import { useState, useEffect } from 'react';
import { CSVLink } from 'react-csv';
import { db } from '../firebase-config';


export const csvexport = () => {
    const [data, setData] = useState<any[]>([]);
    useEffect(() => {
        getDocs(collection(db, "Rooms")).then((docs) => {
            const healthData: any = [];
            docs.forEach((doc) => {
                const {
                    bloodPressure,
                    heartRate,
                    oxygenLevel,
                } = doc.data();
                const userData = {
                    id: doc.id,
                    bloodPressure: bloodPressure,
                    heartRate: heartRate,
                    oxygenLevel: oxygenLevel,
                  };
      
        const headers = [
            { label: "ID", key: data.id },
            { label: "bloodPressure", key: data.bloodPressure },
            { label: "heartRate", key: data.heartRate },
            { label: "oxygenLevel", key: data.oxygenLevel },

        ];

        const csvReport = {
            filename: "userReport.csv",
            headers: headers,
            data: userData
          }
          healthData.push(csvReport);
        });
        setData(healthData);
      });
  }, []);
  
  return (
    <CSVLink  {...data} >
      Export
    </CSVLink>
  )
  }




