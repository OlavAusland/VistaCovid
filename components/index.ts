export * from './Admin'
export * from './ErrorModal'
export * from './Export'
export * from './Home'
export * from './Login'
export * from './Profile'
export * from './Notification'
export * from './Register'
export * from './Room'
