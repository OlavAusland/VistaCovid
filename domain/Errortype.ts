
export type ErrorType = {
    errorObject: Error | undefined;
    errormodalVisible: boolean;
}
