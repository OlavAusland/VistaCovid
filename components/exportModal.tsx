import { Text, View, Pressable, Modal } from 'react-native';
import { useEffect, useState } from 'react';
import Checkbox from "expo-checkbox";

import { exportStyles } from '../styles/ExportModalStyle';
import { ScrollView } from 'react-native-gesture-handler';
import { getOccupiedRooms } from '../api/firebaseAPI';
import { csvexport } from '../utils/csvexport';
import { PreviewModal } from './previewModal';


type exportModalProps = {
    handleRequestClose: Function;
}

export const Exportmodal = (props: exportModalProps) => {
    const [checkedRooms, setCheckedRooms] = useState(new Map());
    const [previevVisible, setPreviewVisible] = useState(false);
    const [csv, setCsv] = useState<string>("");

    useEffect(() => {
        const getRooms = async () => {
            const fetchedRooms = await getOccupiedRooms();
            const checkedBoxes = new Map<string, boolean>();

            fetchedRooms.forEach((item) => {
                checkedBoxes.set(item.id, false);
            });

            setCheckedRooms(checkedBoxes);
        };
        getRooms();
    }, []);


    const handleCheckbox = (key:any, isChecked: any) => {
        setCheckedRooms(new Map(checkedRooms.set(key, isChecked)));
    }

    const handleExport = () => {
        const roomsToExport:string[] = [];
        checkedRooms.forEach((value, key) => {
            if (value) {
                roomsToExport.push(key);
            }
        });
        console.log(roomsToExport);
        handleexportcsv(roomsToExport);
        setPreviewVisible(true);
        
    }

    const handleexportcsv = async(roomsToExport: string[]) => {
        console.log('rooms:', roomsToExport);
             const responce = await csvexport({rooms: roomsToExport});
                setCsv(responce);
    }

    if (previevVisible) {
        return(<PreviewModal csv={csv} setPreviewVisible={setPreviewVisible}/>)
    }


    return (
        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            onRequestClose={() => props.handleRequestClose()}
            testID="patientInfoModal"
        >
            <View style={exportStyles.container}>
                <View style={exportStyles.header}>
                    <Text style={exportStyles.headerText}>Export data to file</Text>
                </View>
                <View style={exportStyles.body}>
                    <View style={exportStyles.bodyRoom}>
                        <Text style={exportStyles.bodyText}>
                            Choose rooms to export
                        </Text>
                        <View style={exportStyles.bodyScroll}>
                            <ScrollView>
                            {checkedRooms.size > 0 &&
                                Array.from(checkedRooms).map(([key, value], index) => {
                                    return (
                                        
                                        <View key={'En key' + index} style={exportStyles.bodyScrollcontent}>
                                            <Checkbox key={"checkbox-" + key} value={value} onValueChange={(isChecked) => handleCheckbox(key, isChecked) } />
                                            <Text style={exportStyles.bodyScrollText}>{key}</Text>
                                           
                                        </View>
                                    );
                                })

                            }
                            </ScrollView>
                        </View>
                    </View>
                    <View style={exportStyles.bodyDate}>
                        <Text style={exportStyles.bodyText}>
                            Select date range
                        </Text>
                    </View>
                    
                </View>
                <View style={exportStyles.footer}>
                <Pressable
                    onPress={() => handleExport()}
                    style={{ flex: 1 }} >
                    <View style={exportStyles.button}>
                        <Text style={exportStyles.footerText}>Export</Text>
                    </View>
                </Pressable>
                <Pressable
                    onPress={() => props.handleRequestClose()}
                    style={{ flex: 1 }} >
                    <View style={exportStyles.button}>
                        <Text style={exportStyles.footerText}>Close</Text>
                    </View>
                </Pressable>
                </View>
                
            </View>
        </Modal>
    );
}
