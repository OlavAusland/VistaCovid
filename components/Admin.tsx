import { View, Text,  ScrollView, SafeAreaView } from 'react-native';
import { adminStyle } from '../styles/AdminStyles';
import Icon from 'react-native-vector-icons/AntDesign';
import Icon3 from 'react-native-vector-icons/MaterialIcons';


// navigation
import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { StackParameters } from '../domain/NavigationTypes';
import { TouchableOpacity } from 'react-native-gesture-handler';

export function AdminView() {

    const navigation = useNavigation<NativeStackNavigationProp<StackParameters>>();


    return(
        <SafeAreaView style={{flex:1}}>
            <ScrollView contentContainerStyle={{flex:1}}>
                <View style={adminStyle.body}>
                    <View style={adminStyle.header}>
                        <Text style={adminStyle.headertext}>ADMIN</Text>
                    </View>
                    <View style={{flex:8}}>
                        <View style={adminStyle.mainViewButtons}>
                            <TouchableOpacity
                            style={adminStyle.TOmain}
                            onPress={()=>{navigation.navigate('ManageRoom')}}>
                                <Icon3 name='meeting-room' size={30} style={adminStyle.buttonIcon} ></Icon3>
                                <Text style={adminStyle.buttonText1}>MANAGE ROOMS</Text>
                                <Icon name='arrowright' size={30} style={adminStyle.arrowIcon}></Icon>
                            </TouchableOpacity>
                        </View>
                        <View style={adminStyle.mainViewButtons}>
                            <TouchableOpacity
                            style={adminStyle.TOmain}
                            onPress={()=>{navigation.navigate('Register')}}>
                                <Icon name='adduser' size={30} style={adminStyle.buttonIcon} ></Icon>
                                <Text style={adminStyle.buttonText2}>REGISTER USER</Text>
                                <Icon name='arrowright' size={30} style={adminStyle.arrowIcon}></Icon>
                            </TouchableOpacity>
                        </View>
                    </View>
                </View>
            </ScrollView>
        </SafeAreaView>

    );
}
