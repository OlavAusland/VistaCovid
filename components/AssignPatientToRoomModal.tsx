import { Text, View, TextInput, Pressable, Modal, TouchableOpacity } from 'react-native';
import React, { useEffect, useState } from 'react';
import { getPatient } from '../api/folkeregisterModelAPI';
import { FolkeregisterPerson } from '../domain/PatientType';
import { assignPatientStyle } from '../styles/AssignPatientStyle';
import { dropdownStyles } from '../styles/dropdownStyle';
import { getAvailableRooms } from '../api/firebaseAPI';
import { Room } from '../domain/RoomType';
import { Dropdown } from 'react-native-element-dropdown';
import FontAwesome5 from 'react-native-vector-icons/FontAwesome5';
import {addPatientToRoom} from '../api/firebaseAPI';



type AssignPatientModalProps = {
    modalVisible: boolean;
    handleRequestClose: Function;
}

type DropDownType = {
    open: boolean,
    value: string,
    items: Array<ItemType>
    room: string
}

type ItemType = {
    label: string,
    value: string
}

export const AssignPatientModal = (props: AssignPatientModalProps) => {

    const [patient, setPatient] = useState<FolkeregisterPerson>();
    const [search, setSearch] = useState<string>("");
    const [dropdown, setDropdown] = useState<DropDownType>({ open: false, value: "0", items: [], room: "" });

    const handleSearch = () => {
        if (search.length > 0) {
            const fnrRegex = new RegExp(/^(0[1-9]|[1-2][0-9]|31(?!(?:0[2469]|11))|30(?!02))(0[1-9]|1[0-2])\d{7}$/g);
            const isFnr = fnrRegex.test(search);

            if (isFnr) {
                getPatient(search).then(result => {
                    setPatient(result)
                }).catch(err => { console.log('here',err.message) });
            }
        }
    }

    const handleAddPatient = () => {
        if (patient && dropdown.room) {
            addPatientToRoom(dropdown.room, patient.ssn);
            props.handleRequestClose();
            setPatient(undefined);
            setSearch("");
            setDropdown({ open: false, value: "0", items: [], room: "" });
        }
    }


    useEffect(() => {
        console.log('here')
        getAvailableRooms().then((room: Room[]) => {
            setDropdown(prev => ({ ...prev, items: [] }));
            room.forEach((room: Room) => {
                const item: ItemType = { label: room.roomNumber, value: room.id, };
                setDropdown(prev => ({ ...prev, items: [...prev.items, item] }));
            })
        })
    }, [props.modalVisible]);

   

    return (

        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.modalVisible}
            onRequestClose={() => props.handleRequestClose()}
        >
            <View style={{ top: '5%', }}>
                <View style={assignPatientStyle.container}>
                    <Text style={{ fontSize: 40, marginBottom: 90 }}>Admit patient</Text>
                    <Text style={{ fontSize: 20, paddingBottom: 5, marginTop: 40 }}>Patient:</Text>
                    <View style={{ width: '100%', borderRadius: 5, flexDirection: 'row', marginBottom: 10 }}>
                        <TextInput onChangeText={text => { setSearch(text) }} placeholder="SSN" style={{ flex: 3, height: 40, backgroundColor: "white", paddingLeft: 10, width: '100%' }} />
                        <TouchableOpacity style={{ flex: 1 }} onPress={() => handleSearch()}>
                            <View style={{ width: 70, backgroundColor: '#0274a1', height: 40, borderRadius: 10, marginLeft: 10 }}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 15, marginTop: 10 }}>Search</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                    {patient &&
                        <View>
                            <Text style={{ fontSize: 15 }}>{patient.lastname}, {patient.firstname} {patient.midlename}</Text>
                        </View>}
                    <Text style={{ fontSize: 20, paddingBottom: 5, marginTop: 25 }}>Room:</Text>
                    <View style={{ backgroundColor: 'white' }}>
                        <Dropdown
                            style={dropdownStyles.dropdown}
                            placeholderStyle={dropdownStyles.placeholderStyle}
                            selectedTextStyle={dropdownStyles.selectedTextStyle}
                            inputSearchStyle={dropdownStyles.inputSearchStyle}
                            iconStyle={dropdownStyles.iconStyle}
                            data={dropdown.items}
                            search
                            maxHeight={200}
                            labelField="label"
                            valueField="value"
                            placeholder="Select room"
                            searchPlaceholder="Search..."
                            value={dropdown.room}
                            onChange={item => {
                                setDropdown(prev => ({ ...prev, room: item.value }));
                            }}
                            renderLeftIcon={() => (
                                <FontAwesome5 style={dropdownStyles.icon} color="black" name="bed" size={20} />
                            )}
                        />
                    </View>
                    <View style={{ flexDirection: 'row', alignContent: 'space-between', marginTop: 200 }}>
                        <Pressable onPress={() => handleAddPatient()} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Add</Text>
                            </View>
                        </Pressable>
                        <TouchableOpacity
                            onPress={() => props.handleRequestClose()} >
                            <View style={assignPatientStyle.button}>
                                <Text style={{ color: 'white', alignSelf: 'center', fontSize: 20 }}>Cancel</Text>
                            </View>
                        </TouchableOpacity>
                    </View>
                </View>
            </View>
        </Modal>
    );
}

