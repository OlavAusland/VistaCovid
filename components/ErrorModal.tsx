import { Modal, View, Text, Pressable } from "react-native";
import { ErrorType } from "../domain/Errortype";
import { erroStyle } from "../styles/ErrorStyle";


type ErrorModalProps = {
    error: ErrorType
    handleRequestClose: Function
   
};

export const Errormodal = (props: ErrorModalProps) => {
    return (
        <Modal
            animationType="slide"
            statusBarTranslucent={true}
            transparent={true}
            visible={props.error.errormodalVisible}
            onRequestClose={() => props.handleRequestClose()}
            testID="patientInfoModal"
        >
            <View style={erroStyle.centeredView}>
                <View style={erroStyle.modalView}>
                    <Text testID="errormessage" style={erroStyle.modalText}>Error: {"\n"}{props.error.errorObject?.message}</Text>
                    
                    <View style={{flex:1}}>
                        <Pressable
                            style={[erroStyle.button, erroStyle.buttonClose]}
                            onPress={() => props.handleRequestClose()} >
                            <Text style={erroStyle.textStyle}>Close</Text>
                        </Pressable>
                    </View>
                </View>
            </View>

        </Modal>
    );
}