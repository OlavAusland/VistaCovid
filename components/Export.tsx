import { Text, View, Pressable, Modal,ScrollView, TouchableOpacity, Touchable  } from 'react-native';
import { createRef, useEffect, useState } from 'react';
import Checkbox from "expo-checkbox";
import { exportStyles } from '../styles/ExportModalStyle';
import { getOccupiedRooms } from '../api/firebaseAPI';
import { csvexport } from '../utils/csvexport';
import { PreviewModal } from './export/previewModal';
import { useNavigation } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { StackParameters } from '../domain/NavigationTypes';
import DateChooser from './export/DateChooser';
import { Moment } from 'moment';
import { styles } from 'react-native-element-dropdown/src/components/TextInput/styles';

type DateState = {
    fromDate: Moment | undefined,
    toDate: Moment | undefined
    showDatePicker: "from" | "to" | undefined
}

export const Export = () => {
    const navigation = useNavigation<NativeStackNavigationProp<StackParameters>>();

    const [checkedRooms, setCheckedRooms] = useState(new Map());
    const [previewVisible, setPreviewVisible] = useState(false);
    const [csv, setCsv] = useState<string>("");
    const [date, setDate] = useState<DateState>({
        fromDate: undefined,
        toDate: undefined,
        showDatePicker: undefined
    });

useEffect(() => {
    const getRooms = async () => {
        const fetchedRooms = await getOccupiedRooms();
        const checkedBoxes = new Map<string, boolean>();

        fetchedRooms.forEach((item) => {
            checkedBoxes.set(item.id, false);
        });

        setCheckedRooms(checkedBoxes);
    };
    getRooms();
}, []);


const handleCheckbox = (key: any) => {
    setCheckedRooms(new Map(checkedRooms.set(key, !checkedRooms.get(key))));
}

const handleExport = async () => {
    const roomsToExport: string[] = [];
    checkedRooms.forEach((value, key) => {
        if (value) {
            roomsToExport.push(key);
        }
    });
 
    const response = await csvexport(roomsToExport, date.fromDate, date.toDate);
    setCsv(response);
    setPreviewVisible(true);

}

const setDateTime = (dateTime: Moment) => {
    if (date.showDatePicker === "from") {
        setDate({
            ...date, fromDate: dateTime, showDatePicker: undefined
        });
    } else if (date.showDatePicker === "to") {
        setDate({
            ...date, toDate: dateTime, showDatePicker: undefined
        });
    }
}

const handleCloseDateChooser = () => {
    setDate({
        ...date, showDatePicker: undefined
    });
}

const dateChooser = () => {
    if (date.showDatePicker === "from") {
        return (
            <DateChooser setDateTime={setDateTime} handleCloseDateChooser={handleCloseDateChooser} />
        );
    } else if (date.showDatePicker === "to") {
        return (
            <DateChooser setDateTime={setDateTime} handleCloseDateChooser={handleCloseDateChooser} />
        );
    }
}

if (previewVisible) {
    return (<PreviewModal csv={csv} setPreviewVisible={setPreviewVisible} />)
}

return(
    <View style={{flex:1}}>
        {dateChooser()}
        <View style={[exportStyles.shadow, {flex:2, justifyContent:'center', alignItems:'center', backgroundColor:'#9DD4FB'}]}> 
            <Text style={{fontSize:35, fontWeight:'bold'}}>Export Data To File</Text>
            <Text style={{fontSize:20}}>Choose Rooms To Export:</Text>
        </View>
        <View style={{flex:6, }}>
            <ScrollView style={[{}]}>     
                {checkedRooms.size > 0 &&
                    Array.from(checkedRooms).map(([key, value], index) => {
                        return (
                            <View key={'En key' + index} style={exportStyles.bodyScrollcontent}>
                                <TouchableOpacity onPress={() => {handleCheckbox(key)}}
                                style={{flexDirection:'row', justifyContent:'center', alignItems:'center'}}>
                                    <Checkbox key={"checkbox-" + key} value={value} onValueChange={() => handleCheckbox(key)} />
                                    <Text style={exportStyles.bodyScrollText}>{key}</Text>
                                </TouchableOpacity>

                            </View>
                        );
                    })
                }           
            </ScrollView>
        </View>
        <View style={[exportStyles.shadow, {flex:1, backgroundColor:'#9DD4FB', paddingTop:5}]}>
            <View style={{flex:1, justifyContent:'center'}}>
                <View style={{flexDirection:'row', justifyContent:'space-around', alignItems:'center'}}>
                    <TouchableOpacity onPress={() => setDate({ ...date, showDatePicker: "from"})} 
                    style={{flexBasis:'45%', justifyContent:'center', alignItems:'center', backgroundColor:'#0274a1', paddingTop:10, paddingBottom:10, borderRadius:10}}>
                        <Text style={{color:'white'}}>Start Date</Text>
                    </TouchableOpacity>
                    <TouchableOpacity onPress={() => setDate({ ...date, showDatePicker: "to"})} 
                    style={{flexBasis:'45%', justifyContent:'center', alignItems:'center', backgroundColor:'#0274a1', paddingTop:10, paddingBottom:10, borderRadius:10}}>
                        <Text style={{color:'white'}}>End Date</Text>
                    </TouchableOpacity>
                </View>
            </View>
            <View style={[{flex:1, justifyContent:'center', alignItems:'center'}]}>
                <TouchableOpacity
                onPress={() => handleExport()}
                style={[{backgroundColor:'#0274a1', width:'95%', alignItems:'center', justifyContent:'center', paddingTop:10, paddingBottom:10, borderRadius:10}]}>
                    <Text style={{color:'white'}}>Export</Text>
                </TouchableOpacity>
            </View>
        </View>

    </View>
)
}
