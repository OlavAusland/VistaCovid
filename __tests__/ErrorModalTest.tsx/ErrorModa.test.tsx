import { act, render } from '@testing-library/react-native';
import { Errormodal } from '../../components';
import { ErrorType } from '../../domain/Errortype';

const errormock: ErrorType = {
    errorObject: new Error('test error'),
    errormodalVisible: true
}
test("renders correctly", () => {
    const tree = render(<Errormodal error={errormock} handleRequestClose ={jest.fn()} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("shows error message", async () => {
    const { getByTestId } = render(<Errormodal error={errormock} handleRequestClose ={jest.fn()} />);
    expect(getByTestId("patientInfoModal")).not.toBeNull();

    const errormessageTest = getByTestId("errormessage")
    await act(async () => {
        expect(errormessageTest.props.children[2]).toContain("test error");
    });
});
