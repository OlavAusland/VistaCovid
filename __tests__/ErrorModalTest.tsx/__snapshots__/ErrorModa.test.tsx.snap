// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  animationType="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  statusBarTranslucent={true}
  testID="patientInfoModal"
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "marginTop": 22,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 20,
          "elevation": 5,
          "height": "40%",
          "margin": 20,
          "padding": 35,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 4,
          "width": "80%",
        }
      }
    >
      <Text
        style={
          Object {
            "flex": 5,
            "marginBottom": 15,
          }
        }
        testID="errormessage"
      >
        Error: 
        

        test error
      </Text>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "borderRadius": 20,
                "elevation": 2,
                "padding": 10,
              },
              Object {
                "backgroundColor": "#2196F3",
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Close
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
